asynction/security.py

```
@@ -0,0 +1,401 @@
+                )
+
+            return token_info, None
+
+        return http_bearer_security_check
+    else:
+        return None
+
+
+def get_cookie_value(cookies: str, name: str) -> Optional[str]:
+    """
+    Returns cookie value by its name. None if no such value.
+    :param cookies: str: cookies raw data
+    :param name: str: cookies key
+
+    Borrowed from https://github.com/zalando/connexion/blob/main/connexion/security/security_handler_factory.py#L206  # noqa: 501
+    """
+    cookie_parser: SimpleCookie = SimpleCookie()
+    cookie_parser.load(str(cookies))
+    try:
+        return cookie_parser[name].value
+    except KeyError:
+        return None
+
+
+def build_http_api_key_security_check(
+    requirement: InternalSecurityRequirement, scheme: SecurityScheme
+) -> Optional[InternalSecurityCheck]:
+    api_key_info_func = load_api_key_info_func(scheme)
+    scope_validate_func = load_scope_validate_func(scheme)
+
+    _, required_scopes = requirement
+    api_key_in = scheme.in_
+    api_key_name = scheme.name
+    if api_key_name is None:
+        raise SecurityException("invalid api key name specified")
+    if api_key_in not in ["query", "header", "cookie"]:
+        raise SecurityException("invalid api key location specified")
+
+    def http_api_key_security_check(request: Request) -> InternalSecurityCheckResponse:
+        api_key = None
+
+        if api_key_in == "query":
+            if api_key_name is not None:
+                api_key = request.args.get(api_key_name)
+        elif api_key_in == "header":
+            if api_key_name is not None:
+                api_key = request.headers.get(api_key_name)
+        elif api_key_in == "cookie":
+            cookies_list = request.headers.get("Cookie")
+            if cookies_list and api_key_name is not None:
+                api_key = get_cookie_value(cookies_list, api_key_name)
+        else:
+            return None, None
+
+        if api_key is None:
+            return None, None
```

Not sure why we always return `None` as the second element of the tuple. Is this expected to be an error message? If we always return it as None, then we probably don't need it.

---

Maybe we can just raise an exception rather than returning `None, None`, and remove the `if err` logic?

---

Let's drop the `err` and raise an exception whenever things go wrong. I remember advising you otherwise, but I think this combination of raising SecurityException and returning `None, None` is turning out to be confusing.